### ISSUE TITLE
Poor Generalization Performance

### ISSUE THREAD

**Issue by alokwarey** [dynamicslab/pysindy#108]
I fit a continuous time model to experimental data - one feature and 3 control variables. The model does very well on the training data. However, when I use the model to simulate a test case it does poorly. If the model is doing well on the training data then it has learned the dynamics of the system. I am unsure why then it performs poorly on a new test case? The range of control variables for the test case are covered in the training data. Even with a reduced feature library to avoid overfitting, it still does poorly on the test set.

**TRAIN**
![MRT](https://user-images.githubusercontent.com/69644995/118868892-bd9ffb80-b8b2-11eb-9fd6-564f2b1c50d0.png)

**TEST
![MRT2](https://user-images.githubusercontent.com/69644995/118868928-c7c1fa00-b8b2-11eb-9c50-f71a1dbc2f54.png)
**

**Comment by akaptano** [comment]
Hi alokwarey,

Please post your fitted model for this so we can see the explicit equations. If there is an analytic solution, please post that as well. Please also post your call to the SINDy fit method... for instance, it is unclear what you are using for the SINDy hyperparameters like the threshold, regularizer choice (L0, L1, ...), and so on. 

I'm not sure I would agree that the trained model is fit very well. There seem to be four features: an initial spike, a second peak, a little square wave, and an overall background shape to the plot (the SINDy model only approximates this last feature). Without additional information, I also disagree that "if the model is doing well on the training data then it has learned the dynamics of the system," even if the control variable ranges are covered in the training data. It may still be overfitting or otherwise specific to the particular training trajectory. 

Please post the additional information and I will try to help you figure out how to improve the model.

Best,
Alan Kaptanoglu

**Comment by alokwarey** [comment]
Hi Alan,

Thanks so much for the response. Those features are noise/errors in the experimental data. Our goal is to get a model that captures the general features of the problem. We don't have an analytical model for this complex system. 

I agree with you that training with one trajectory is not ideal. The trajectory we picked covered the range of control variables that we typically encounter for this problem. However, I agree, we should train with multiple trajectories. 

**Details of the model:**

differentiation_method = ps.SmoothedFiniteDifference()
feature_library = ps.PolynomialLibrary(degree=2, include_interaction=False, include_bias=True)
optimizer = ps.STLSQ(threshold=1e-3)
model = ps.SINDy(differentiation_method=differentiation_method,
    feature_library=feature_library,
    optimizer=optimizer,    
)

model.fit(x=x_train, u=u_train_scaled, t=dt) 

dt == 1s
Control variables are scaled using standard scaler. Not sure if this is needed?


SINDy(differentiation_method=SmoothedFiniteDifference(),
      feature_library=PolynomialLibrary(include_interaction=False),
      feature_names=['x0', 'x1', 'u0', 'u1', 'u2'],
      optimizer=STLSQ(threshold=0.001))

x0' = -0.030 1 + -0.001 x0 + 0.002 x1 + 0.017 u0 + -0.009 u2 + 0.010 u0^2
x1' = 0.146 1 + 0.006 x0 + -0.011 x1 + 0.037 u0 + -0.003 u0^2 + -0.002 u1^2 + -0.006 u2^2

The plots in my previous post are for x0.

**Comment by akaptano** [comment]
Great, thanks for the information. A few things to try/think about:
1. If the small-scale features I mentioned are not of interest, or are known errors, I would recommend smoothing these features out. 
2. Could you post what the model looks like for a sparser model? What about a nonsparse model (lambda = 0)?
3. My impression is the default regularizer for STLSQ is L1. What happens when you use L0?
4. I actually don't know about the scaling of the control variables... you might need to ask Brian about this. I have not made a model using control variables before. 

Best,
Alan

**Comment by alokwarey** [comment]
I did search through the hyperparameter space. Will try a few additional options. Thanks!

**Comment by briandesilva** [comment]
Thanks, Alan, for looking into this. On the topic of rescaling control variables, rescaling will usually help, though sometimes it is unnecessary. Though unsatisfying, I think the best advice I can give you is to try both and see which gives you better results.

It looks like the training and testing datasets are a bit noisy, so I'd also recommend trying out a few differentiation methods with smoothing. I see that you have already tried `SmoothedFiniteDifference`, but it might help to test out some others like splines. See the examples [here](https://pysindy.readthedocs.io/en/latest/examples/5_differentation.html).
